### PR TITLE
Support PPE profiling for `Local`

### PR DESCRIPTION
**Pull request by KantaTamura**
This PR resolves the `Local` task in #258 

> [!NOTE]
> To enable tracing, `trace=True` must be added during `Local()` initialization (or `from_url()`).

Sample program to check tracing
```python
import json
import pytorch_pfn_extras as ppe

from pfio.v2 import Local, Path, from_url

@ppe.profiler.record_function("my_tag_2", trace=True)
def my_run2(x):
    pass

def my_run1(x):
    pass

def p():
    tracer = ppe.profiler.get_tracer()

    tracer.clear()
    with ppe.profiler.record("my_tag_1", trace=True):
        my_run1('hoge')

    for i in range(4):
        my_run2("far")

    dir = 0
    fil = 0
    with Local(trace=True) as fs:
        for f in fs.list():
            if fs.isdir(f.strip()):
                dir += 1
                continue
            
            fil += 1
            len(fs.open(f).read())

    print(fil, "files", dir, "dirs")

    dict = tracer.state_dict()
    keys = [event["name"] for event in json.loads(dict["_event_list"])]
    print(keys)

    w = ppe.writing.SimpleWriter(out_dir="")

    # output '['
    tracer.initialize_writer("trace.json", w)
    # output json dump
    tracer.flush("trace.json", w)

p()
```

output
```
9 files 15 dirs
['my_tag_1', 'my_tag_2', 'my_tag_2', 'my_tag_2', 'my_tag_2', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-0', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-1', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-2', 'pfio.v2.Local:list-3', 'pfio.v2.Local:list-4', 'pfio.v2.Local:list-5', 'pfio.v2.Local:list-6', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-7', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-8', 'pfio.v2.Local:list-9', 'pfio.v2.Local:list-10', 'pfio.v2.Local:list-11', 'pfio.v2.Local:list-12', 'pfio.v2.Local:list-13', 'pfio.v2.Local:list-14', 'pfio.v2.Local:list-15', 'pfio.v2.Local:list-16', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-17', 'pfio.v2.Local:list-18', 'pfio.v2.Local:list-19', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-20', 'pfio.v2.Local:list-21', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-22', 'pfio.v2.Local:open', 'pfio.v2.Local:read', 'pfio.v2.Local:list-23']
```

drawing of the output json file (`trace.json`) with `chrome://tracing` is shown below
image

### CODE REVIEW

**Review by k5342 — COMMENTED** [review]
Overall LGTM. I have one minor suggestion.

**Comment by k5342** [comment]
@kuenishi Could you take a look? Please let us know if you have any comments on the tag naming rule and the granularity of the tests.

File: setup.py [hunk 1]
```diff
@@ -52,7 +52,7 @@
     packages=find_packages(),
     package_data={'pfio': package_data},
     extras_require={
-        'test': ['pytest', 'flake8', 'autopep8', 'parameterized', 'isort', 'moto[server]>=5.0.0', 'numpy', 'mypy'],
```
[the remark this answers is not in the document]
Comment (kuenishi) [reply setup.py:55]: Agree. We'd rather add something like `trace` than adding the dependency to `test`. PFIO users shall be using profiling without this extra requires and will be installing PFIO with `pip install pfio` . In that case, PPE won't be installed even though they need it.

File: tests/v2_tests/test_local.py [hunk 2]
```diff
@@ -291,6 +293,45 @@ def test_from_url_create_option(self):
         from_url(path, create=True)
         assert os.path.exists(path) and os.path.isdir(path)
 
+    def test_ppe_profiling(self):
```
Comment (kuenishi) [inline tests/v2_tests/test_local.py:296]: Please don't extend the `TestLocal(unittest.TestCase)` class. We're slowly moving on from the `unitest` style, to a more pytest-like style where each test is defined as a single and independent function (sometimes with fixtures or parameterized). Please refer to `test_s3.py`.

**Review by kuenishi — CHANGES_REQUESTED** [review]
The Python implementation itself looks good, but I have a few request to align with current style and options.

File: pfio/v2/local.py [hunk 3]
```diff
@@ -4,9 +4,54 @@
 import shutil
 from typing import Optional
 
+try:
```
Comment (kuenishi) [inline pfio/v2/local.py:7]: Refer to: https://github.com/pfnet/pfio/pull/326#discussion_r1329411782

Comment (KantaTamura) [reply pfio/v2/local.py]: Split into `pfio/v2/_profiler.py` and changed to import from `_profiler.py`.

File: tests/v2_tests/test_local.py (hunk 2 above)
Comment (KantaTamura) [reply tests/v2_tests/test_local.py]: Changed to not use `TestLocal(unittest.TestCase)`.

Currently, `foo.txt` is generated in the directory where pytest runs, do you want to use a different path? (e.g. `/tmp/...`)

File: setup.py [hunk 4]
```diff
@@ -52,11 +52,12 @@
...
+        'test': ['pytest', 'flake8', 'autopep8', 'parameterized', 'isort', 'moto[server]>=5.0.0', 'numpy', 'mypy', 'pytorch-pfn-extras'],
```
Comment (kuenishi) [inline setup.py:55]: As we discussed offline, PPE shouldn't be here but tests must be switched in the test code using something like [skipif](https://docs.pytest.org/en/stable/how-to/skipping.html#id1).

Comment (KantaTamura) [reply setup.py]: Changed to using `importorskip()` to skip if no ppe.

- no ppe
```
test_local.py ................s
```

- exist ppe
```
test_local.py .................
```

Comment (k5342) [reply setup.py]: We may also need to tweak the tox.ini to enable CI testing for the tracing code, which we're adding this PR if we remove PPE from the test section in setup.py: https://github.com/pfnet/pfio/blob/b874d493abb80ca7836f439fb7af37bac327b182/tox.ini#L4-L20

Comment (kuenishi) [reply setup.py]: Good catch

**Review by kuenishi — APPROVED** [review]
Please leave the summary of offline discussion with PPE developer why we don't need complete JSON file and it works fine without it.

**Review by kuenishi — CHANGES_REQUESTED** [review]
Ah, and don't forget to update tox.ini to include `trace` option as @k5342 pointed out.

**Comment by KantaTamura** [comment]
Offline discussion summary: Chrome Tracing does not require valid json and can be rendered with just `initialize_writer()` and `flush()` as in the [sample program](https://github.com/pfnet/pfio/pull/340#issue-2479646572).

ref. [Trace Event Format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.yr4qxyxotyw)
> When provided as a string to the importer the ] at the end of the JSON Array Format is optional.

**Comment by KantaTamura** [comment]
fix `tox.ini` and summarized the json output method in the documentation (tips).

**Review by kuenishi — APPROVED** [review]

**Pull request merged**